utils/s3/client.cc

```diff
@@ -214,46 +225,104 @@ inline bool is_redirect_status(http::reply::status_type st) {
+        case aws::aws_error_type::ACCESS_DENIED:
+            error_code = EACCES;
+            break;
+        default:
+            error_code = EIO;
+        }
+        return make_ready_future<storage_io_error>(storage_io_error(error_code, format("S3 request failed. Reason: {}", ex.what())));
     } catch (const httpd::unexpected_status_error& e) {
         auto status = e.status();
 
         if (is_redirect_status(status) || status == http::reply::status_type::not_found) {
-            return make_exception_future<>(storage_io_error(ENOENT, format("S3 object doesn't exist ({})", status)));
+            return make_ready_future<storage_io_error>(storage_io_error(ENOENT, format("S3 object doesn't exist ({})", status)));
         }
         if (status == http::reply::status_type::forbidden || status == http::reply::status_type::unauthorized) {
-            return make_exception_future<>(storage_io_error(EACCES, format("S3 access denied ({})", status)));
+            return make_ready_future<storage_io_error>(storage_io_error(EACCES, format("S3 access denied ({})", status)));
         }
 
-        return make_exception_future<>(storage_io_error(EIO, format("S3 request failed with ({})", status)));
+        return make_ready_future<storage_io_error>(storage_io_error(EIO, format("S3 request failed with ({})", status)));
     } catch (...) {
         auto e = std::current_exception();
-        return make_exception_future<>(storage_io_error(EIO, format("S3 error ({})", e)));
+        return make_ready_future<storage_io_error>(storage_io_error(EIO, format("S3 error ({})", e)));
     }
+}
+
+future<> client::do_retryable_request(group_client& gc, http::request req, http::experimental::client::reply_handler handler) const {
+    uint32_t retries = 0;
+    do {
+        std::exception_ptr e;
+        aws::aws_exception request_ex{aws::aws_error{}};
+        try {
+            co_return co_await gc.http.make_request(req, handler, std::nullopt);
+        } catch (const aws::aws_exception& ex) {
+            ++retries;
+            e = std::current_exception();
+            request_ex = ex;
+        }
+        if (_retry_strategy->should_retry(request_ex.error(), retries)) {
+            s3l.debug("S3 client request failed. Reason: {}. Retry# {}", request_ex.what(), retries);
+            co_await seastar::sleep(_retry_strategy->delay_before_retry(request_ex.error(), retries));
+        } else {
+            if (!request_ex.error().is_retryable()) {
+                s3l.error("S3 client encountered non-retryable error. Reason: {}. Code: {}. Retry# {}", request_ex.what(),
+                        std::to_underlying(request_ex.error().get_error_type()), retries);
+            } else {
+                s3l.error("Retries exhausted. Retry# {}", retries);
+            }
+            throw co_await map_s3_client_exception(e);
+        }
+    } while (true);
+}
 
+http::experimental::client::reply_handler client::make_s3_error_handler(
```

Comment:
Can CompleteMultipartUpload provide its own error handler that checks response body regardless of the received status? It's not very nice that this generic helper is equipped with specific "I'm called by CompleteMultipartUpload" parameter :(

Reply:
how about now?

Reply:
Now it looks awesome. Please also consider the following. Here's how the code looks now:

```c++
future<> client::make_request(http::request req,
                              custom_error_handler error_handler,
                              http::experimental::client::reply_handler handle,
                              http::reply::status_type expected) {
    authorize(req);
    auto& gc = find_or_create_client();
    auto custom_handler = error_handler(std::move(handle), expected);

    return do_retryable_request(gc, std::move(req), std::move(custom_handler));
}
```

the `custom_error_handler` is callable (std::function) that returns another function. And note, that this callable is called instantly here, so why not just directly pass the result of its invocation into this make_request()?